### PR TITLE
Giza: introduce complex distribution bucket IDs.

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
This PR modifies the `DistributionBucketId` from the simple u64 type to the structure based composed of `DistributionBucketFamilyId` and `DistributionBucketIndex`. 

Changes:
- reworked complex DistributionBucketId
- added DistributionBucketIndex type and changed DistributionBucketFamily type
- changed the random bucket selection algorithm for the distribution buckets.
- `types` library was changed

It supersedes PRs - https://github.com/Joystream/joystream/pull/2797 and https://github.com/Joystream/joystream/pull/2792

Relates to https://github.com/Joystream/joystream/issues/2865

### CODE REVIEW

File: types/src/storage.ts [hunk 1]
```diff
@@ -220,13 +232,23 @@ export class DistributionBucket
...
 
+export type IDistributionBucketState = {
+  accepting_new_bags: bool
+}
+
+export class DistributionBucketState
+  extends JoyStructDecorated({
+    accepting_new_bags: bool,
+  })
+  implements IDistributionBucketState {}
```
Comment (Lezek123) [inline types/src/storage.ts]: I couldn't find this type defined anywhere in the runtime

**Review by Lezek123 — CHANGES_REQUESTED** [review]

File: types/src/storage.ts (hunk 1 above)
Comment (shamil-gadelshin) [reply types/src/storage.ts]: This is dead code. I removed it here 14c4633

File: runtime-modules/storage/src/lib.rs [hunk 2]
```diff
@@ -826,45 +828,42 @@ impl<Balance: Saturating + Copy> BagUpdate<Balance> {
...
 
-    // Checks inner buckets for bag assignment number. Returns true only if all 'assigned_bags' are
-    // zero.
-    fn no_bags_assigned(&self) -> bool {
-        self.distribution_buckets
-            .values()
-            .all(|b| b.no_bags_assigned())
-    }
+/// Type alias for the DistributionBucketIdRecord.
+pub type DistributionBucketId<T> = DistributionBucketIdRecord<
```
Comment (bedeho) [inline runtime-modules/storage/src/lib.rs:851]: Is it risky, or confusing, to have `DistributionBucketId` and `DistributionBucketIndex` denote distinct but intimately related concepts? This caused me a bit of confusion when reading through.

Perhaps the latter could be something like `WithinFamilyDistributionBucketId` or something? just a tip, you decide what you want to do.

File: runtime-modules/storage/src/random_buckets/storage_bucket_picker.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+pub(crate) use super::{RandomBucketIdIterator, SequentialBucketIdIterator};
+use crate::{DynamicBagType, Module, Trait};
+
+// Generates storage bucket IDs to assign to a new dynamic bag.
+pub(crate) struct StorageBucketPicker<T> {
+    trait_marker: PhantomData<T>,
+}
+
+impl<T: Trait> StorageBucketPicker<T> {
+    // Selects storage bucket ID sets to assign to the storage bucket.
```
Comment (bedeho) [inline runtime-modules/storage/src/random_buckets/storage_bucket_picker.rs]: `assign to the bag`

**Review by bedeho — APPROVED** [review]
This looks good to me, approving under the assumption that there will be fix later which introduces the ability, and perhaps replaces wholesale the current permissive mode, to create dynamic bags which are guaranteed to be stored and distributed in a policy compliant way.

File: runtime-modules/storage/src/random_buckets/storage_bucket_picker.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/src/random_buckets/storage_bucket_picker.rs]: Fixed.

File: runtime-modules/storage/src/lib.rs (hunk 2 above)
Comment (shamil-gadelshin) [reply runtime-modules/storage/src/lib.rs:851]: `DistributionBucketIndex` seemed to satisfy the rest of the team. So, I tend to keep the name for now.

**Pull request merged**